src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs

```diff
@@ -0,0 +1,527 @@
+        {
+            InternalWriteItemNull();
+            BinaryArrayTypeEnum binaryArrayTypeEnum;
+            var lengthA = new int[1];
+            lengthA[0] = length;
+            int[] lowerBoundA = null;
+            object typeInformation = null;
+
+            if (lowerBound == 0)
+            {
+                binaryArrayTypeEnum = BinaryArrayTypeEnum.Single;
+            }
+            else
+            {
+                binaryArrayTypeEnum = BinaryArrayTypeEnum.SingleOffset;
+                lowerBoundA = new int[1];
+                lowerBoundA[0] = lowerBound;
+            }
+
+            int assemId;
+            BinaryTypeEnum binaryTypeEnum = BinaryTypeConverter.GetBinaryTypeInfo(
+                arrayElemTypeNameInfo._NItype, objectInfo, arrayElemTypeNameInfo.NIname, _objectWriter, out typeInformation, out assemId);
+
+            if (_binaryArray == null)
+            {
+                _binaryArray = new BinaryArray();
+            }
+            _binaryArray.Set((int)arrayNameInfo._NIobjectId, 1, lengthA, lowerBoundA, binaryTypeEnum, typeInformation, binaryArrayTypeEnum, assemId);
+
+            _binaryArray.Write(this);
+
+            if (Converter.IsWriteAsByteArray(arrayElemTypeNameInfo._NIprimitiveTypeEnum) && (lowerBound == 0))
+            {
+                //array is written out as an array of bytes
+                if (arrayElemTypeNameInfo._NIprimitiveTypeEnum == InternalPrimitiveTypeE.Byte)
+                {
+                    WriteBytes((byte[])array);
+                }
+                else if (arrayElemTypeNameInfo._NIprimitiveTypeEnum == InternalPrimitiveTypeE.Char)
+                {
+                    WriteChars((char[])array);
+                }
+                else
+                {
+                    WriteArrayAsBytes(array, Converter.TypeLength(arrayElemTypeNameInfo._NIprimitiveTypeEnum));
+                }
+            }
+        }
+
+        private void WriteArrayAsBytes(Array array, int typeLength)
+        {
+            InternalWriteItemNull();
+            int byteLength = array.Length * typeLength;
+            int arrayOffset = 0;
+            if (_byteBuffer == null)
+            {
+                _byteBuffer = new byte[ChunkSize];
+            }
+
+            while (arrayOffset < array.Length)
+            {
+                int numArrayItems = Math.Min(ChunkSize / typeLength, array.Length - arrayOffset);
+                int bufferUsed = numArrayItems * typeLength;
+                Buffer.BlockCopy(array, arrayOffset * typeLength, _byteBuffer, 0, bufferUsed);
+                if (!BitConverter.IsLittleEndian)
+                {
+                    // we know that we are writing a primitive type, so just do a simple swap
+                    for (int i = 0; i < bufferUsed; i += typeLength)
```

Comment:
> How much, if any, faith do we have in this block?

Probably not much.  In desktop this is under an ifdef.  I could delete the block entirely and just assert IsLittleEndian?  Or leave the code but add a Debug.Fail to the block to alert someone to re-review it?